Add public API for deleting costumes and sounds

paulkaplan opened the pull request:
This provides methods on the public VM object to delete sounds and costumes on the editing target. These will be used by the GUI instead of reaching into the internal vm data.

paulkaplan committed:
Add delete costume and sound functions to the vm public api

paulkaplan committed:
Add to the vm spec

rschamp commented on src/virtual-machine.js:
```diff
@@ -284,6 +284,20 @@ class VirtualMachine extends EventEmitter {
+            .concat(this.editingTarget.sprite.costumes.slice(costumeIndex + 1));
+        if (costumeIndex === this.editingTarget.currentCostume) {
+            this.editingTarget.setCostume(costumeIndex - 1);
+        }
```
Would you please add some tests? Curious about these cases:
* `costumeIndex === currentCostume === 0`: Would expect currentCostume to remain 0, but I believe it will wrap to the last costume
* `costumeIndex === 0`, `currentCostume === 2`, `costumes.length === 3`: Would expect currentCostume to decrease to maintain the same costume, but I believe it will wrap to 0
* `costumes.length === 1`: We should probably reject attempts to delete the last costume?

rschamp replied:
And jumping off of my comment on https://github.com/LLK/scratch-vm/pull/568, can we move this logic to the `RenderedTarget`? The target should maintain its own asset lists.

paulkaplan replied:
Will do. I'll just lean on scratch2 behavior for those test cases.

paulkaplan committed:
Make rendered targets in charge of deleting their own sounds and costumes

paulkaplan commented:
@rschamp I paired with @fsih on e0e8235, added tests for deleting sounds and costumes, which should address your questions. We also moved the responsibility of deleting into rendered target.

rschamp commented on src/sprites/rendered-target.js:
```diff
@@ -390,6 +390,39 @@ class RenderedTarget extends Target {
+     * @param {number} index Sound index to be deleted
+     */
+    deleteSound (index) {
+        if (this.sprite.sounds.length === 1) return;
```
I think deleting the last sound would be ok — sprites don't require a sound to work.

rschamp commented on src/sprites/rendered-target.js:
```diff
@@ -390,6 +390,39 @@ class RenderedTarget extends Target {
+            this.setCostume(index - 1);
+        } else {
+            this.setCostume(index);
+        }
```
This still seems to do an odd thing - say you have 5 costumes:
```
Costume 1
Costume 2
Costume 3
Costume 4   <-- current costume
Costume 5

Costume 1
-Costume 2- <-- deleted costume
Costume 3
Costume 4   <-- current costume
Costume 5

Behavior                           Expected
Costume 1                          Costume 1                      
Costume 3   <-- current costume    Costume 3                      
Costume 4                          Costume 4   <-- current costume
Costume 5                          Costume 5                      
```
If we allow deleting costumes other than the currently selected one, this is weird behavior — we should maintain the current costume if we delete an index < than the current costume index.

rschamp reviewed (changes requested):
Thank you for refactoring that into `RenderedTarget` and adding tests. I think we need to consider when costumes.length > 3 though.

paulkaplan committed:
Allow deleting the last sound

paulkaplan committed:
Handle more cases of deleting costumes

paulkaplan commented:
@rschamp Added a few more test cases, used that ascii art syntax to describe them because I ran out of words. Good catch.

rschamp reviewed (changes requested):
When I delete the current costume, the change is not reflected on the renderer? It's not immediately apparent how that would have regressed with this code though (this behavior isn't present on the current build). To repro: load the default project and delete the first costume. The renderer also seems to be painting — you can use move blocks and the sprite will move still using the old deleted costume.

paulkaplan committed:
Always activate the current costume after deleting

paulkaplan commented:
@rschamp good catch, the problem was that there was a branch of the `if` that didn't call `setCostume` because the current costume index doesn't change. however, the _costume_ that index was referring to changed, so I needed to add an extra `else` to the bottom to make sure `setCostume` is always called.

rschamp reviewed (approved):
Looks good!
